packages/xsnap/src/xsnap.js

```diff
@@ -340,35 +411,112 @@ export async function xsnap(options) {
     try {
-      const result = baton.then(async () => {
-        await messagesToXsnap.next(encoder.encode(`w${tmpSnapPath}`));
-        await runToIdle();
+      /** @type {string} */
+      let snapPath;
+      /** @type {Readable} */
+      let sourceStream;
+
+      const maybePipe = () => {
+        if (!piped && !cleaned) {
+          sourceStream.pipe(output);
+          piped = true;
+        }
+      };
+
+      if (snapshotUseFs) {
+        // TODO: Refactor to use tmpFile rather than tmpName.
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        snapPath = await ptmpName({
+          template: `make-snapshot-${safeHintFromDescription(
+            description,
+          )}-XXXXXX.xss`,
+        });
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        const handle = await fs.open(snapPath, 'w+');
+        // @ts-expect-error 'close' event added in Node 15.4
+        handle.on('close', () => {
+          fs.unlink(snapPath);
+        });
+        sourceStream = handle.createReadStream();
+        finished(output).finally(() => sourceStream.destroy());
+      } else {
+        sourceStream = savedSnapshotsStream;
+        snapPath = '@7';
+        batonKit.promise.then(maybePipe);
+      }
+
+      const cleanup = () => {
+        if (cleaned) return;
+        cleaned = true;
+        sourceStream.unpipe(output);
+        // eslint-disable-next-line no-use-before-define
+        output.off('data', onData);
+        output.end();
+      };
+      const checkDone = () => {
+        if (snapshotSize !== undefined && snapshotReadSize >= snapshotSize) {
+          cleanup();
+        }
+      };
+      const onData = chunk => {
+        snapshotReadSize += chunk.length;
+        checkDone();
+      };
+      output.on('data', onData);
+
+      const result = batonKit.promise.then(async () => {
+        await messagesToXsnap.next(encoder.encode(`w${snapPath}`));
+        const { reply } = await runToIdle();
+        const lengthStr = decoder.decode(reply);
+        if (lengthStr.length) {
+          snapshotSize = Number(lengthStr);
+          maybePipe();
+        } else {
+          snapshotSize = -1;
+        }
+        checkDone();
       });
-      baton = result.then(noop, noop);
-      // eslint-disable-next-line @jessie.js/no-nested-await
-      await racePromises([vatExit.promise, baton]);
-      const snapReader = fs.createReadStream(tmpSnapPath);
-      // eslint-disable-next-line @jessie.js/no-nested-await
-      await fsStreamReady(snapReader);
-      yield* snapReader;
+      batonKit.resolve(result);
+      done = racePromises([vatCancelled, result]);
+      done.catch(() => cleanup());
+
+      yield* output;
     } finally {
       // eslint-disable-next-line @jessie.js/no-nested-await
-      await fs.unlink(tmpSnapPath);
+      await done;
+      (piped && snapshotReadSize === snapshotSize) ||
+        Fail`Snapshot size does not match. saved=${q(snapshotSize)}, read=${q(
```

Comment:
ah, ok, here's one place that would signal an error after `snapshotSize = -1`, and since this at the top of `async function* makeSnapshotInternal`, the generator given to snapStore will exit with an error. I think that works.